File: 14.py
def apply_mask(val, mask):
    val = bin(val).lstrip("0b").zfill(len(mask))
    masked_val = []
    for x, m in zip(val, mask):
        if m == "X":
            masked_val.append(x)
        else:
            masked_val.append(m)
    return int("".join(masked_val), 2)


def apply_mask2(val, mask):
    val = bin(val).lstrip("0b").zfill(len(mask))
    masked_val = []
    floating = []
    for i, (x, m) in enumerate(zip(val, mask)):
        if m == "0":
            masked_val.append(x)
            continue

        if m == "X":
            floating.append(i)

        masked_val.append(m)

    for i in range(2 ** len(floating)):
        x = bin(i).lstrip("0b").zfill(len(floating))
        for i, b in zip(floating, x):
            masked_val[i] = b

        yield int("".join(masked_val), 2)


if __name__ == "__main__":
    with open("14.in") as f:
        input = [l.strip() for l in f.readlines()]

    # Part 1
    mem = {}
    mask = ""
    for l in input:
        op, val = l.split(" = ")
        if op == "mask":
            mask = val
            continue

        addr = int(op.lstrip("mem[").rstrip("]"))
        mem[addr] = apply_mask(int(val), mask)

    print(sum(mem.values()))

    # Part 2
    mem = {}
    mask = ""
    for l in input:
        op, val = l.split(" = ")
        if op == "mask":
            mask = val
            continue

        addr = int(op.lstrip("mem[").rstrip("]"))
        for a in apply_mask2(int(addr), mask):
            mem[a] = int(val)

    print(sum(mem.values()))

File: 14.go
package main

import (
	"fmt"
	"io/ioutil"
	"math"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func applyMask1(val int, mask string) int {
	binary := strconv.FormatInt(int64(val), 2)
	padded := fmt.Sprintf("%0*s", len(mask), binary)
	masked := make([]byte, len(padded))
	for i := 0; i < len(masked); i++ {
		if mask[i] == 'X' {
			masked[i] = padded[i]
		} else {
			masked[i] = mask[i]
		}
	}
	result, _ := strconv.ParseInt(string(masked), 2, 64)
	return int(result)
}

func applyMask2(val int, mask string) []int {
	binary := strconv.FormatInt(int64(val), 2)
	padded := fmt.Sprintf("%0*s", len(mask), binary)
	masked := make([]byte, len(padded))
	floating := []int{}
	for i := 0; i < len(masked); i++ {
		if mask[i] == '0' {
			masked[i] = padded[i]
		} else if mask[i] == 'X' {
			floating = append(floating, i)
		} else {
			masked[i] = mask[i]
		}
	}

	addrs := []int{}
	for i := 0; i < int(math.Pow(2, float64(len(floating)))); i++ {
		binary := strconv.FormatInt(int64(i), 2)
		padded := fmt.Sprintf("%0*s", len(floating), binary)
		for i, _ := range padded {
			masked[floating[i]] = padded[i]
		}
		result, _ := strconv.ParseInt(string(masked), 2, 64)
		addrs = append(addrs, int(result))
	}
	return addrs
}

func part1(lines []string) int {
	maskRe := regexp.MustCompile(`^mask = ([01X]+)$`)
	memRe := regexp.MustCompile(`^mem\[(\d+)\] = (\d+)$`)
	mem := make(map[int]int)
	var mask string
	for _, l := range lines {
		if maskMatch := maskRe.FindStringSubmatch(l); len(maskMatch) > 0 {
			mask = maskMatch[1]
			continue
		}

		if memMatch := memRe.FindStringSubmatch(l); len(memMatch) > 0 {
			addr, _ := strconv.Atoi(memMatch[1])
			val, _ := strconv.Atoi(memMatch[2])
			mem[addr] = applyMask1(val, mask)
		}
	}

	var sum int
	for _, v := range mem {
		sum += v
	}
	return sum
}

func part2(lines []string) int {
	maskRe := regexp.MustCompile(`^mask = ([01X]+)$`)
	memRe := regexp.MustCompile(`^mem\[(\d+)\] = (\d+)$`)
	mem := make(map[int]int)
	var mask string
	for _, l := range lines {
		if maskMatch := maskRe.FindStringSubmatch(l); len(maskMatch) > 0 {
			mask = maskMatch[1]
			continue
		}

		if memMatch := memRe.FindStringSubmatch(l); len(memMatch) > 0 {
			addr, _ := strconv.Atoi(memMatch[1])
			val, _ := strconv.Atoi(memMatch[2])
			for _, a := range applyMask2(addr, mask) {
				mem[a] = val
			}
		}
	}

	var sum int
	for _, v := range mem {
		sum += v
	}
	return sum
}

func main() {
	bytes, err := ioutil.ReadFile("14.in")
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	lines := strings.Split(strings.TrimSpace(string(bytes)), "\n")

	fmt.Println(part1(lines))
	fmt.Println(part2(lines))
}
